File: .claude/skills/execute-testcases-from-excel/scripts/excel_io.js
#!/usr/bin/env node
/*
 * excel_io.js — Bản Node tương đương excel_io.py cho skill `execute-testcases-from-excel`.
 * Dùng khi môi trường KHÔNG có Python (chỉ có Node). Cần package `exceljs`.
 *
 *   node excel_io.js read  <input.xlsx|.csv> [--sheet NAME] [--out run/testcases.json]
 *   node excel_io.js write <input.xlsx> --results run/results.json [--out out.xlsx] [--in-place] [--sheet NAME]
 *
 * exceljs được resolve qua NODE_PATH nếu chưa cài cục bộ, ví dụ:
 *   NODE_PATH=<scratchpad>/tcgen/node_modules node excel_io.js read ...
 */
'use strict';
const fs = require('fs');
const path = require('path');
let ExcelJS;
try { ExcelJS = require('exceljs'); }
catch (e) {
  console.error('[excel_io] Thiếu package exceljs. Cài: npm i exceljs (hoặc set NODE_PATH tới node_modules có exceljs).');
  process.exit(2);
}

const FIELD_ALIASES = {
  tc_id:       ['tc id','tcid','tc','ma tc','ma test case','ma testcase','id','test case id','test id','ma','stt tc'],
  module:      ['module','chuc nang','mo dun','phan he','nhom chuc nang'],
  risk_level:  ['risk level','risk','muc rui ro','do rui ro','muc do rui ro'],
  title:       ['test title','title','tieu de','ten test case','ten testcase','test scenario','scenario','kich ban','test case','mo ta','description','noi dung','ten'],
  precondition:['pre condition','precondition','tien dieu kien','dieu kien tien quyet','dieu kien'],
  steps:       ['test steps','steps','cac buoc','cac buoc thuc hien','buoc thuc hien','steps to reproduce','thao tac','cac buoc kiem thu'],
  expected:    ['expected result','expected','ket qua mong doi','ket qua ky vong','ket qua mong muon','mong doi','ket qua du kien'],
  priority:    ['priority','do uu tien','muc uu tien','uu tien'],
  test_data:   ['test data','du lieu test','du lieu','data','du lieu kiem thu'],
};
const RESULT_COLUMNS = {
  'Actual Result': ['actual result','ket qua thuc te','actual','thuc te'],
  'Status':        ['status','trang thai','ket qua chay','verdict'],
  'Executed At':   ['executed at','thoi gian chay','ngay chay','execution time'],
  'Evidence':      ['evidence','bang chung','screenshot','anh','hinh anh'],
  'Note / Bug ID': ['note / bug id','note','bug id','ghi chu','bug','note bug id'],
};
const RESULT_ORDER = Object.keys(RESULT_COLUMNS);
const STATUS_FILL = { PASS:'FFC6EFCE', FAIL:'FFFFC7CE', BLOCKED:'FFD9D9D9', SKIP:'FFFFEB9C' };
const STATUS_FONT = { PASS:'FF006100', FAIL:'FF9C0006', BLOCKED:'FF3F3F3F', SKIP:'FF9C6500' };

function norm(s) {
  if (s === null || s === undefined) return '';
  s = String(s).trim().toLowerCase().replace(/đ/g, 'd');
  s = s.normalize('NFD').replace(/\p{Mn}/gu, '');
  for (const ch of ['_','-','.',':','(',')','/','\\','*','#']) s = s.split(ch).join(' ');
  return s.split(/\s+/).filter(Boolean).join(' ');
}
function matchField(cell) {
  const h = norm(cell);
  if (!h) return null;
  for (const [f, aliases] of Object.entries(FIELD_ALIASES)) if (aliases.includes(h)) return f;
  for (const [f, aliases] of Object.entries(FIELD_ALIASES)) for (const a of aliases) if (a.length >= 3 && h.includes(a)) return f;
  return null;
}
function detectHeader(rows, scan = 15) {
  let bestIdx = null, bestMap = {};
  for (let i = 0; i < Math.min(rows.length, scan); i++) {
    const mapping = {};
    const used = new Set();
    rows[i].forEach((cell, j) => {
      const f = matchField(cell);
      if (f && !used.has(f)) { mapping[j] = f; used.add(f); }
    });
    if (Object.keys(mapping).length > Object.keys(bestMap).length) { bestIdx = i; bestMap = mapping; }
  }
  if (bestIdx === null || Object.keys(bestMap).length < 2) {
    console.error('[excel_io] Không nhận diện được header test case (cần >=2 cột khớp). Kiểm tra file có TC ID, Test Steps, Expected Result... hoặc --sheet đúng.');
    process.exit(1);
  }
  return [bestIdx, bestMap];
}
function cellText(v) {
  if (v === null || v === undefined) return '';
  if (typeof v === 'object') {
    if (Array.isArray(v.richText)) return v.richText.map(t => t.text).join('');
    if ('text' in v) return String(v.text);
    if ('result' in v) return String(v.result);
    if ('formula' in v) return String(v.result != null ? v.result : '');
    if (v instanceof Date) return v.toISOString();
  }
  return String(v);
}
async function loadRowsXlsx(input, sheet) {
  const wb = new ExcelJS.Workbook();
  await wb.xlsx.readFile(input);
  const ws = sheet ? wb.getWorksheet(sheet) : wb.worksheets[0];
  if (!ws) { console.error('[excel_io] Không tìm thấy sheet: ' + sheet); process.exit(1); }
  const rows = [];
  const rc = ws.rowCount, cc = ws.columnCount;
  for (let r = 1; r <= rc; r++) {
    const row = ws.getRow(r), arr = [];
    for (let c = 1; c <= cc; c++) arr.push(cellText(row.getCell(c).value));
    rows.push(arr);
  }
  return [rows, sheet || ws.name, wb, ws];
}
function loadRowsCsv(input) {
  const txt = fs.readFileSync(input, 'utf8').replace(/^﻿/, '');
  const rows = [];
  let field = '', row = [], inq = false;
  for (let i = 0; i < txt.length; i++) {
    const ch = txt[i];
    if (inq) {
      if (ch === '"' && txt[i+1] === '"') { field += '"'; i++; }
      else if (ch === '"') inq = false; else field += ch;
    } else if (ch === '"') inq = true;
    else if (ch === ',') { row.push(field); field = ''; }
    else if (ch === '\n') { row.push(field); rows.push(row); field = ''; row = []; }
    else if (ch === '\r') { /* skip */ }
    else field += ch;
  }
  if (field.length || row.length) { row.push(field); rows.push(row); }
  return [rows, 'csv'];
}

async function cmdRead(args) {
  const ext = path.extname(args.input).toLowerCase();
  let rows, sheet;
  if (ext === '.xlsx' || ext === '.xlsm') { [rows, sheet] = await loadRowsXlsx(args.input, args.sheet); }
  else if (ext === '.csv') { [rows, sheet] = loadRowsCsv(args.input); }
  else { console.error('[excel_io] Định dạng không hỗ trợ: ' + ext + ' (dùng .xlsx hoặc .csv).'); process.exit(1); }

  const [headerIdx, colmap] = detectHeader(rows);
  const records = [];
  for (let i = headerIdx + 1; i < rows.length; i++) {
    const row = rows[i];
    const rec = {}; for (const k of Object.keys(FIELD_ALIASES)) rec[k] = '';
    for (const [j, field] of Object.entries(colmap)) rec[field] = (row[j] != null ? String(row[j]).trim() : '');
    if (!Object.values(rec).some(v => String(v).trim())) continue;
    rec._row = i + 1; rec._sheet = sheet;
    records.push(rec);
  }
  const colsDetected = {}; for (const [j, f] of Object.entries(colmap)) colsDetected[f] = Number(j);
  const out = {
    source: path.resolve(args.input), sheet, header_row: headerIdx + 1,
    columns_detected: colsDetected, count: records.length, testcases: records,
  };
  const payload = JSON.stringify(out, null, 2);
  if (args.out) {
    fs.mkdirSync(path.dirname(path.resolve(args.out)), { recursive: true });
    fs.writeFileSync(args.out, payload, 'utf8');
    console.log(`[excel_io] Đọc ${records.length} test case -> ${args.out}`);
    console.log(`[excel_io] Cột nhận diện: ${Object.values(colmap).sort().join(', ')}`);
  } else console.log(payload);
}

function loadResults(p) {
  const data = JSON.parse(fs.readFileSync(p, 'utf8'));
  const items = Array.isArray(data) ? data : (data.results || []);
  const byRow = {}, byId = {};
  for (const it of items) {
    if (it._row != null) byRow[parseInt(it._row, 10)] = it;
    if (it.tc_id) byId[norm(it.tc_id)] = it;
  }
  return [byRow, byId];
}
function resultValue(item, key) {
  const ev = item.evidence;
  switch (key) {
    case 'Actual Result': return item.actual || '';
    case 'Status': return String(item.status || '').trim().toUpperCase();
    case 'Executed At': return item.executed_at || new Date().toISOString().slice(0,19).replace('T',' ');
    case 'Evidence': return Array.isArray(ev) ? ev.join(', ') : (ev || '');
    case 'Note / Bug ID': return [item.bug_id || '', item.note || ''].filter(Boolean).join(' ').trim();
    default: return '';
  }
}
async function cmdWrite(args) {
  const ext = path.extname(args.input).toLowerCase();
  if (ext !== '.xlsx' && ext !== '.xlsm') { console.error('[excel_io] Bản Node chỉ ghi .xlsx (CSV: dùng bản py).'); process.exit(1); }
  const [byRow, byId] = loadResults(args.results);
  const wb = new ExcelJS.Workbook();
  await wb.xlsx.readFile(args.input);
  const ws = args.sheet ? wb.getWorksheet(args.sheet) : wb.worksheets[0];
  const rc = ws.rowCount, cc = ws.columnCount;
  const rows = [];
  for (let r = 1; r <= rc; r++) { const row = ws.getRow(r), arr = []; for (let c = 1; c <= cc; c++) arr.push(cellText(row.getCell(c).value)); rows.push(arr); }
  const [headerIdx, colmap] = detectHeader(rows);
  const headerRowNo = headerIdx + 1;
  const tcidCol = Object.entries(colmap).find(([, f]) => f === 'tc_id');
  const tcidJ = tcidCol ? Number(tcidCol[0]) : null;

  const existing = {};
  rows[headerIdx].forEach((cell, j) => {
    const hn = norm(cell);
    for (const [canon, aliases] of Object.entries(RESULT_COLUMNS))
      if (hn === norm(canon) || aliases.includes(hn)) existing[canon] = j;
  });
  let nextCol = rows[headerIdx].length;
  const resultCol = {};
  for (const canon of RESULT_ORDER) {
    if (canon in existing) resultCol[canon] = existing[canon];
    else { resultCol[canon] = nextCol; ws.getRow(headerRowNo).getCell(nextCol + 1).value = canon; nextCol++; }
  }
  let written = 0;
  for (let i = headerIdx + 1; i < rows.length; i++) {
    const excelRow = i + 1;
    let item = byRow[excelRow];
    if (!item && tcidJ != null && tcidJ < rows[i].length) item = byId[norm(rows[i][tcidJ])];
    if (!item) continue;
    const status = String(item.status || '').trim().toUpperCase();
    for (const [canon, jcol] of Object.entries(resultCol))
      ws.getRow(excelRow).getCell(jcol + 1).value = resultValue(item, canon);
    if (STATUS_FILL[status]) {
      const scell = ws.getRow(excelRow).getCell(resultCol['Status'] + 1);
      scell.fill = { type: 'pattern', pattern: 'solid', fgColor: { argb: STATUS_FILL[status] } };
      scell.font = { color: { argb: STATUS_FONT[status] }, bold: (status === 'FAIL' || status === 'PASS') };
    }
    written++;
  }
  let out = args.in_place ? args.input : (args.out || args.input.replace(/(\.xlsm|\.xlsx)$/i, '_executed$1'));
  fs.mkdirSync(path.dirname(path.resolve(out)), { recursive: true });
  await wb.xlsx.writeFile(out);
  console.log(`[excel_io] Ghi ${written} kết quả -> ${out}`);
}

function parseArgs(argv) {
  const a = { _: [] };
  for (let i = 0; i < argv.length; i++) {
    const t = argv[i];
    if (t === '--in-place') a.in_place = true;
    else if (t === '--sheet') a.sheet = argv[++i];
    else if (t === '--out') a.out = argv[++i];
    else if (t === '--results') a.results = argv[++i];
    else a._.push(t);
  }
  return a;
}
(async () => {
  const [cmd, ...rest] = process.argv.slice(2);
  const args = parseArgs(rest);
  args.input = args._[0];
  if (!cmd || !args.input) { console.error('Dùng: node excel_io.js <read|write> <file> [--sheet] [--out] [--results] [--in-place]'); process.exit(1); }
  if (cmd === 'read') await cmdRead(args);
  else if (cmd === 'write') { if (!args.results) { console.error('[excel_io] write cần --results'); process.exit(1); } await cmdWrite(args); }
  else { console.error('[excel_io] Lệnh không hợp lệ: ' + cmd); process.exit(1); }
})().catch(e => { console.error('[excel_io] Lỗi:', e.message); process.exit(1); });

File: .claude/skills/execute-testcases-from-excel/scripts/excel_io.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
excel_io.py — Helper cho skill `execute-testcases-from-excel`.

Chức năng:
  read  : Đọc test case từ Excel (.xlsx) hoặc CSV -> JSON chuẩn hóa.
          Tự nhận diện dòng header + map cột theo alias (Tiếng Việt & English).
  write : Ghi kết quả thực thi (từ results.json) ngược vào file, thêm/điền các cột
          Actual Result / Status / Executed At / Evidence / Note-Bug ID và tô màu
          theo trạng thái (chỉ với .xlsx).

Ví dụ:
  python excel_io.py read  testcases.xlsx --sheet "Sheet1" --out run/testcases.json
  python excel_io.py write testcases.xlsx --results run/results.json --out testcases_executed.xlsx
  python excel_io.py write testcases.xlsx --results run/results.json --in-place

Ghi chú: .xls (định dạng cũ) không hỗ trợ trực tiếp — hãy lưu lại thành .xlsx.
"""
import argparse
import csv
import json
import os
import sys
import unicodedata
from datetime import datetime

# ----------------------------------------------------------------------------
# Map cột nguồn (đọc test case) — key chuẩn -> danh sách alias (đã bỏ dấu)
# ----------------------------------------------------------------------------
FIELD_ALIASES = {
    "tc_id":       ["tc id", "tcid", "tc", "ma tc", "ma test case", "ma testcase",
                    "id", "test case id", "test id", "ma", "stt tc"],
    "module":      ["module", "chuc nang", "mo dun", "phan he", "nhom chuc nang"],
    "risk_level":  ["risk level", "risk", "muc rui ro", "do rui ro", "muc do rui ro"],
    "title":       ["test title", "title", "tieu de", "ten test case", "ten testcase",
                    "test case", "mo ta", "description", "noi dung", "ten"],
    "precondition":["pre condition", "precondition", "tien dieu kien",
                    "dieu kien tien quyet", "dieu kien"],
    "steps":       ["test steps", "steps", "cac buoc", "cac buoc thuc hien",
                    "buoc thuc hien", "steps to reproduce", "thao tac", "cac buoc kiem thu"],
    "expected":    ["expected result", "expected", "ket qua mong doi",
                    "ket qua ky vong", "ket qua mong muon", "mong doi", "ket qua du kien"],
    "priority":    ["priority", "do uu tien", "muc uu tien", "uu tien"],
    "test_data":   ["test data", "du lieu test", "du lieu", "data", "du lieu kiem thu"],
}

# Cột kết quả (ghi ngược) — header chuẩn -> alias để phát hiện cột đã tồn tại
RESULT_COLUMNS = {
    "Actual Result": ["actual result", "ket qua thuc te", "actual", "thuc te"],
    "Status":        ["status", "trang thai", "ket qua chay", "verdict"],
    "Executed At":   ["executed at", "thoi gian chay", "ngay chay", "execution time"],
    "Evidence":      ["evidence", "bang chung", "screenshot", "anh", "hinh anh"],
    "Note / Bug ID": ["note / bug id", "note", "bug id", "ghi chu", "bug", "note bug id"],
}
RESULT_ORDER = list(RESULT_COLUMNS.keys())

# Màu nền theo trạng thái (ARGB) cho .xlsx
STATUS_FILL = {
    "PASS":    "FFC6EFCE",
    "FAIL":    "FFFFC7CE",
    "BLOCKED": "FFD9D9D9",
    "SKIP":    "FFFFEB9C",
}
STATUS_FONT = {
    "PASS":    "FF006100",
    "FAIL":    "FF9C0006",
    "BLOCKED": "FF3F3F3F",
    "SKIP":    "FF9C6500",
}


def norm(s):
    """Chuẩn hóa chuỗi để so khớp: thường hóa, bỏ dấu tiếng Việt, gộp khoảng trắng."""
    if s is None:
        return ""
    s = str(s).strip().lower()
    s = s.replace("đ", "d")  # đ -> d (ký tự này KHÔNG tách dấu dưới NFD)
    s = unicodedata.normalize("NFD", s)
    s = "".join(c for c in s if unicodedata.category(c) != "Mn")
    for ch in ["_", "-", ".", ":", "(", ")", "/", "\\", "*", "#"]:
        s = s.replace(ch, " ")
    return " ".join(s.split())


def match_field(header_cell):
    """Trả về key chuẩn nếu header khớp một alias, ngược lại None."""
    h = norm(header_cell)
    if not h:
        return None
    for field, aliases in FIELD_ALIASES.items():
        if h in aliases:
            return field
    # khớp mềm: header chứa alias (cho các header dài như "Các bước thực hiện (Steps)")
    for field, aliases in FIELD_ALIASES.items():
        for a in aliases:
            if a in h and len(a) >= 3:
                return field
    return None


def detect_header(rows, scan=15):
    """Tìm dòng header: dòng trong `scan` dòng đầu map được NHIỀU cột nhất (>=2).
    Trả về (header_index, {col_index: field})."""
    best_idx, best_map = None, {}
    for i, row in enumerate(rows[:scan]):
        mapping = {}
        for j, cell in enumerate(row):
            f = match_field(cell)
            if f and f not in mapping.values():
                mapping[j] = f
        if len(mapping) > len(best_map):
            best_idx, best_map = i, mapping
    if best_idx is None or len(best_map) < 2:
        raise SystemExit(
            "[excel_io] Không nhận diện được header test case (cần >=2 cột khớp).\n"
            "  Kiểm tra file có các cột như: TC ID, Test Steps, Expected Result...\n"
            "  Hoặc chỉ định đúng --sheet."
        )
    return best_idx, best_map


# ----------------------------------------------------------------------------
# Đọc file -> danh sách các dòng (list[list[str]])
# ----------------------------------------------------------------------------
def load_rows_xlsx(path, sheet):
    try:
        from openpyxl import load_workbook
    except ImportError:
        raise SystemExit("[excel_io] Thiếu openpyxl. Cài: pip install openpyxl --break-system-packages")
    wb = load_workbook(path, data_only=True, read_only=True)
    ws = wb[sheet] if sheet else wb[wb.sheetnames[0]]
    rows = []
    for r in ws.iter_rows(values_only=True):
        rows.append(["" if c is None else str(c) for c in r])
    return rows, (sheet or wb.sheetnames[0])


def load_rows_csv(path):
    with open(path, "r", encoding="utf-8-sig", newline="") as f:
        return [list(r) for r in csv.reader(f)], "csv"


def cmd_read(args):
    ext = os.path.splitext(args.input)[1].lower()
    if ext in (".xlsx", ".xlsm"):
        rows, sheet = load_rows_xlsx(args.input, args.sheet)
    elif ext == ".csv":
        rows, sheet = load_rows_csv(args.input)
    else:
        raise SystemExit(f"[excel_io] Định dạng không hỗ trợ: {ext} (dùng .xlsx hoặc .csv; .xls hãy lưu lại thành .xlsx).")

    header_idx, colmap = detect_header(rows)
    records = []
    for i in range(header_idx + 1, len(rows)):
        row = rows[i]
        rec = {k: "" for k in FIELD_ALIASES}
        for j, field in colmap.items():
            rec[field] = (row[j].strip() if j < len(row) and row[j] is not None else "")
        if not any(str(v).strip() for v in rec.values()):
            continue  # bỏ dòng trống
        rec["_row"] = i + 1          # số dòng Excel (1-based)
        rec["_sheet"] = sheet
        records.append(rec)

    out = {
        "source": os.path.abspath(args.input),
        "sheet": sheet,
        "header_row": header_idx + 1,
        "columns_detected": {v: k for k, v in colmap.items()},
        "count": len(records),
        "testcases": records,
    }
    payload = json.dumps(out, ensure_ascii=False, indent=2)
    if args.out:
        os.makedirs(os.path.dirname(os.path.abspath(args.out)), exist_ok=True)
        with open(args.out, "w", encoding="utf-8") as f:
            f.write(payload)
        print(f"[excel_io] Đọc {len(records)} test case -> {args.out}")
        print(f"[excel_io] Cột nhận diện: {', '.join(sorted(colmap.values()))}")
    else:
        print(payload)


# ----------------------------------------------------------------------------
# Ghi kết quả ngược vào .xlsx / .csv
# ----------------------------------------------------------------------------
def load_results(path):
    with open(path, "r", encoding="utf-8") as f:
        data = json.load(f)
    items = data.get("results", data) if isinstance(data, dict) else data
    by_row, by_id = {}, {}
    for it in items:
        if it.get("_row") is not None:
            by_row[int(it["_row"])] = it
        if it.get("tc_id"):
            by_id[norm(it["tc_id"])] = it
    return by_row, by_id


def result_value(item, key):
    ev = item.get("evidence")
    if key == "Actual Result":
        return item.get("actual", "")
    if key == "Status":
        return str(item.get("status", "")).strip().upper()
    if key == "Executed At":
        return item.get("executed_at") or datetime.now().strftime("%Y-%m-%d %H:%M:%S")
    if key == "Evidence":
        return ", ".join(ev) if isinstance(ev, list) else (ev or "")
    if key == "Note / Bug ID":
        note = item.get("note", "") or ""
        bug = item.get("bug_id", "") or ""
        return " ".join(x for x in [bug, note] if x).strip()
    return ""


def cmd_write(args):
    ext = os.path.splitext(args.input)[1].lower()
    by_row, by_id = load_results(args.results)
    if ext in (".xlsx", ".xlsm"):
        _write_xlsx(args, by_row, by_id)
    elif ext == ".csv":
        _write_csv(args, by_row, by_id)
    else:
        raise SystemExit(f"[excel_io] Định dạng không hỗ trợ để ghi: {ext}")


def _resolve_out(args, default_suffix="_executed"):
    if args.in_place:
        return args.input
    if args.out:
        return args.out
    stem, ext = os.path.splitext(args.input)
    return f"{stem}{default_suffix}{ext if ext in ('.xlsx', '.xlsm') else '.xlsx'}"


def _write_xlsx(args, by_row, by_id):
    try:
        from openpyxl import load_workbook
        from openpyxl.styles import PatternFill, Font
    except ImportError:
        raise SystemExit("[excel_io] Thiếu openpyxl. Cài: pip install openpyxl --break-system-packages")

    wb = load_workbook(args.input)
    ws = wb[args.sheet] if args.sheet else wb[wb.sheetnames[0]]
    rows = [[("" if c.value is None else str(c.value)) for c in r] for r in ws.iter_rows()]
    header_idx, colmap = detect_header(rows)
    header_row_no = header_idx + 1

    # Tìm cột tc_id để khớp dự phòng
    tcid_col = next((j for j, f in colmap.items() if f == "tc_id"), None)

    # Xác định / tạo cột kết quả
    existing = {}
    header_cells = rows[header_idx]
    for j, cell in enumerate(header_cells):
        hn = norm(cell)
        for canon, aliases in RESULT_COLUMNS.items():
            if hn == norm(canon) or hn in aliases:
                existing[canon] = j
    next_col = len(header_cells)
    result_col = {}
    for canon in RESULT_ORDER:
        if canon in existing:
            result_col[canon] = existing[canon]
        else:
            result_col[canon] = next_col
            ws.cell(row=header_row_no, column=next_col + 1, value=canon)
            next_col += 1

    written = 0
    for i in range(header_idx + 1, len(rows)):
        excel_row = i + 1
        item = by_row.get(excel_row)
        if item is None and tcid_col is not None and tcid_col < len(rows[i]):
            item = by_id.get(norm(rows[i][tcid_col]))
        if not item:
            continue
        status = str(item.get("status", "")).strip().upper()
        for canon, jcol in result_col.items():
            ws.cell(row=excel_row, column=jcol + 1, value=result_value(item, canon))
        if status in STATUS_FILL:
            fill = PatternFill("solid", fgColor=STATUS_FILL[status])
            font = Font(color=STATUS_FONT[status], bold=(status in ("FAIL", "PASS")))
            scol = result_col["Status"] + 1
            ws.cell(row=excel_row, column=scol).fill = fill
            ws.cell(row=excel_row, column=scol).font = font
        written += 1

    out = _resolve_out(args)
    os.makedirs(os.path.dirname(os.path.abspath(out)), exist_ok=True)
    wb.save(out)
    print(f"[excel_io] Ghi {written} kết quả -> {out}")


def _write_csv(args, by_row, by_id):
    rows, _ = load_rows_csv(args.input)
    header_idx, colmap = detect_header(rows)
    tcid_col = next((j for j, f in colmap.items() if f == "tc_id"), None)
    header = rows[header_idx]
    base = len(header)
    result_col = {c: base + k for k, c in enumerate(RESULT_ORDER)}
    header = header + RESULT_ORDER

    out_rows = rows[:header_idx] + [header]
    written = 0
    for i in range(header_idx + 1, len(rows)):
        row = list(rows[i]) + [""] * len(RESULT_ORDER)
        excel_row = i + 1
        item = by_row.get(excel_row)
        if item is None and tcid_col is not None and tcid_col < len(rows[i]):
            item = by_id.get(norm(rows[i][tcid_col]))
        if item:
            for canon, jcol in result_col.items():
                row[jcol] = result_value(item, canon)
            written += 1
        out_rows.append(row)

    stem, _ = os.path.splitext(args.input)
    out = args.input if args.in_place else (args.out or f"{stem}_executed.csv")
    with open(out, "w", encoding="utf-8-sig", newline="") as f:
        csv.writer(f).writerows(out_rows)
    print(f"[excel_io] Ghi {written} kết quả -> {out}")


def main():
    p = argparse.ArgumentParser(description="Đọc/ghi test case Excel cho skill execute-testcases-from-excel")
    sub = p.add_subparsers(dest="cmd", required=True)

    pr = sub.add_parser("read", help="Đọc Excel/CSV -> JSON test case chuẩn hóa")
    pr.add_argument("input")
    pr.add_argument("--sheet", default=None)
    pr.add_argument("--out", default=None, help="Đường dẫn JSON xuất (bỏ trống = in ra stdout)")
    pr.set_defaults(func=cmd_read)

    pw = sub.add_parser("write", help="Ghi kết quả (results.json) ngược vào file")
    pw.add_argument("input")
    pw.add_argument("--results", required=True, help="File results.json")
    pw.add_argument("--sheet", default=None)
    pw.add_argument("--out", default=None, help="File xuất (mặc định <tên>_executed.xlsx)")
    pw.add_argument("--in-place", action="store_true", help="Ghi đè trực tiếp file gốc")
    pw.set_defaults(func=cmd_write)
